File: domain-services/gofeeler/app/auth/auth_test.go
package auth

import (
	"crypto/rand"
	"crypto/rsa"
	"encoding/base64"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// Real RSA keypair + a real local HTTP server standing in for Keycloak's
// JWKS endpoint, so signature verification is exercised against real
// crypto rather than mocked away entirely.
func generateKeyPair(t *testing.T) *rsa.PrivateKey {
	t.Helper()
	key, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		t.Fatalf("generate RSA key: %v", err)
	}
	return key
}

func startMockJWKS(t *testing.T, kid string, key *rsa.PrivateKey) string {
	t.Helper()
	n := base64.RawURLEncoding.EncodeToString(key.PublicKey.N.Bytes())
	e := base64.RawURLEncoding.EncodeToString(big64(key.PublicKey.E))

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"keys": []map[string]string{
				{"kty": "RSA", "kid": kid, "use": "sig", "alg": "RS256", "n": n, "e": e},
			},
		})
	}))
	t.Cleanup(server.Close)
	return server.URL
}

func big64(e int) []byte {
	// Standard RSA public exponent (65537) fits in 3 bytes — matches
	// how Keycloak's own JWKS encodes "e" ("AQAB").
	b := []byte{byte(e >> 16), byte(e >> 8), byte(e)}
	for len(b) > 1 && b[0] == 0 {
		b = b[1:]
	}
	return b
}

func signToken(t *testing.T, key *rsa.PrivateKey, kid string, expOffset time.Duration, claims *Claims) string {
	t.Helper()
	if claims.ExpiresAt == nil {
		claims.ExpiresAt = jwt.NewNumericDate(time.Now().Add(expOffset))
	}
	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	token.Header["kid"] = kid
	signed, err := token.SignedString(key)
	if err != nil {
		t.Fatalf("sign token: %v", err)
	}
	return signed
}

// Real Keycloak access tokens always carry aud: "account" — included by
// default here so tests exercise the same shape a real token has. A
// sibling Rust implementation of this exact fix (asset-service/
// rustledger) shipped a real bug where the JWT library's default
// validation silently rejected every real token over this exact field,
// caught only because a live test happened to include it — baking it
// into every test claims value here so that class of bug can't hide.
func baseClaims(roles []string) *Claims {
	return &Claims{
		RegisteredClaims: jwt.RegisteredClaims{
			Subject:  "user-1",
			Audience: jwt.ClaimStrings{"account"},
		},
		RealmAccess: &realmAccess{Roles: roles},
	}
}

func TestClaimsFromHeader_ValidTokenVerifies(t *testing.T) {
	key := generateKeyPair(t)
	kid := "test-key-1"
	jwksURLOverride = startMockJWKS(t, kid, key)
	t.Cleanup(func() { jwksURLOverride = ""; resetCache() })

	token := signToken(t, key, kid, 5*time.Minute, baseClaims([]string{"platform:project-manager"}))
	claims := ClaimsFromHeader("Bearer " + token)

	if claims == nil {
		t.Fatal("expected claims, got nil")
	}
	if !claims.HasRole("platform:project-manager") {
		t.Error("expected platform:project-manager role")
	}
	if claims.Subject != "user-1" {
		t.Errorf("expected sub=user-1, got %q", claims.Subject)
	}
}

func TestClaimsFromHeader_MissingHeaderReturnsNil(t *testing.T) {
	if claims := ClaimsFromHeader(""); claims != nil {
		t.Errorf("expected nil, got %+v", claims)
	}
}

func TestClaimsFromHeader_NonBearerSchemeReturnsNil(t *testing.T) {
	if claims := ClaimsFromHeader("Basic dXNlcjpwYXNz"); claims != nil {
		t.Errorf("expected nil, got %+v", claims)
	}
}

func TestClaimsFromHeader_ForgedTokenIsRejected(t *testing.T) {
	key := generateKeyPair(t)
	otherKey := generateKeyPair(t)
	kid := "test-key-2"
	jwksURLOverride = startMockJWKS(t, kid, key)
	t.Cleanup(func() { jwksURLOverride = ""; resetCache() })

	// Signed with a different key than the one the mock JWKS serves
	// under this kid — the actual forged-token scenario this fix closes.
	forged := signToken(t, otherKey, kid, 5*time.Minute, baseClaims([]string{"platform:admin"}))
	if claims := ClaimsFromHeader("Bearer " + forged); claims != nil {
		t.Errorf("expected forged token to be rejected, got %+v", claims)
	}
}

func TestClaimsFromHeader_ExpiredTokenIsRejected(t *testing.T) {
	key := generateKeyPair(t)
	kid := "test-key-3"
	jwksURLOverride = startMockJWKS(t, kid, key)
	t.Cleanup(func() { jwksURLOverride = ""; resetCache() })

	// Well outside any reasonable clock-skew leeway.
	expired := signToken(t, key, kid, -1*time.Hour, baseClaims([]string{"platform:admin"}))
	if claims := ClaimsFromHeader("Bearer " + expired); claims != nil {
		t.Errorf("expected expired token to be rejected, got %+v", claims)
	}
}

func TestClaimsFromHeader_UnrecognizedKidIsRejected(t *testing.T) {
	key := generateKeyPair(t)
	jwksURLOverride = startMockJWKS(t, "the-real-kid", key)
	t.Cleanup(func() { jwksURLOverride = ""; resetCache() })

	token := signToken(t, key, "not-the-real-kid", 5*time.Minute, baseClaims([]string{"platform:admin"}))
	if claims := ClaimsFromHeader("Bearer " + token); claims != nil {
		t.Errorf("expected unrecognized-kid token to be rejected, got %+v", claims)
	}
}

func TestClaimsFromHeader_NoRealmAccessHasNoRoles(t *testing.T) {
	key := generateKeyPair(t)
	kid := "test-key-4"
	jwksURLOverride = startMockJWKS(t, kid, key)
	t.Cleanup(func() { jwksURLOverride = ""; resetCache() })

	claims := &Claims{RegisteredClaims: jwt.RegisteredClaims{Subject: "user-1", Audience: jwt.ClaimStrings{"account"}}}
	token := signToken(t, key, kid, 5*time.Minute, claims)
	got := ClaimsFromHeader("Bearer " + token)

	if got == nil {
		t.Fatal("expected claims, got nil")
	}
	if got.HasRole("platform:admin") {
		t.Error("expected no roles")
	}
}

// resetCache clears the package-level JWKS cache between tests so one
// test's mock server URL/keys can't leak into another's assertions.
func resetCache() {
	cacheMu.Lock()
	cachedKeys = nil
	cacheMu.Unlock()
}

File: platform-services/blog-service/middleware/auth.test.js
const { test, describe, before, after } = require('node:test');
const assert = require('node:assert/strict');
const http = require('node:http');
const crypto = require('node:crypto');
const jwt = require('jsonwebtoken');

const { claimsFromHeader, syncClaims, requireAnyRealmRole, setJwksUriForTests } = require('./auth');

// Real RSA keypair + a tiny local HTTP server standing in for Keycloak's
// JWKS endpoint, so signature verification is exercised against real
// crypto rather than mocked away entirely.
let server;
let privateKey;
const KID = 'test-key-1';

before(async () => {
  const { publicKey, privateKey: priv } = crypto.generateKeyPairSync('rsa', { modulusLength: 2048 });
  privateKey = priv;
  const jwk = publicKey.export({ format: 'jwk' });

  server = http.createServer((req, res) => {
    res.setHeader('Content-Type', 'application/json');
    res.end(JSON.stringify({ keys: [{ ...jwk, kid: KID, use: 'sig', alg: 'RS256' }] }));
  });
  await new Promise((resolve) => server.listen(0, '127.0.0.1', resolve));
  const { port } = server.address();
  setJwksUriForTests(`http://127.0.0.1:${port}/certs`);
});

after(() => {
  server.close();
});

function signToken(payload, overrides = {}) {
  return jwt.sign(payload, privateKey, {
    algorithm: 'RS256',
    keyid: KID,
    expiresIn: '5m',
    ...overrides,
  });
}

function fakeRes() {
  const res = { statusCode: null, body: null };
  res.status = (code) => {
    res.statusCode = code;
    return res;
  };
  res.json = (body) => {
    res.body = body;
    return res;
  };
  return res;
}

describe('claimsFromHeader', () => {
  test('a validly signed token verifies and returns its claims', async () => {
    const token = signToken({ sub: 'user-1', realm_access: { roles: ['platform:marketing'] } });
    const claims = await claimsFromHeader(`Bearer ${token}`);
    assert.deepEqual(claims.realm_access.roles, ['platform:marketing']);
  });

  test('missing header returns null', async () => {
    assert.equal(await claimsFromHeader(undefined), null);
  });

  test('non-Bearer scheme throws', async () => {
    await assert.rejects(() => claimsFromHeader('Basic dXNlcjpwYXNz'));
  });

  test('a token forged with an unknown key is rejected — this is the actual vulnerability closed here', async () => {
    // Before real verification, a forged token claiming
    // platform:marketing would sail past requireAnyRealmRole below,
    // since only the role claim was ever checked, never its signature.
    const { privateKey: otherKey } = crypto.generateKeyPairSync('rsa', { modulusLength: 2048 });
    const forged = jwt.sign({ realm_access: { roles: ['platform:marketing'] } }, otherKey, {
      algorithm: 'RS256',
      keyid: KID,
      expiresIn: '5m',
    });
    await assert.rejects(() => claimsFromHeader(`Bearer ${forged}`));
  });

  test('an expired token is rejected', async () => {
    const token = signToken({ sub: 'user-1' }, { expiresIn: '-10s' });
    await assert.rejects(() => claimsFromHeader(`Bearer ${token}`));
  });
});

describe('syncClaims', () => {
  test('no Authorization header: claims stay null, request proceeds anonymously', async () => {
    const req = { headers: {} };
    const res = fakeRes();
    let nextCalled = false;
    await syncClaims(req, res, () => {
      nextCalled = true;
    });
    assert.equal(nextCalled, true);
    assert.equal(req.claims, null);
  });

  test('a present but invalid token is rejected with 401', async () => {
    const req = { headers: { authorization: 'Bearer garbage' } };
    const res = fakeRes();
    let nextCalled = false;
    await syncClaims(req, res, () => {
      nextCalled = true;
    });
    assert.equal(nextCalled, false);
    assert.equal(res.statusCode, 401);
  });
});

describe('requireAnyRealmRole', () => {
  test('calls next when the caller holds at least one of the roles', () => {
    const req = { claims: { realm_access: { roles: ['platform:marketing'] } } };
    const res = fakeRes();
    let nextCalled = false;
    requireAnyRealmRole('platform:admin', 'platform:marketing')(req, res, () => {
      nextCalled = true;
    });
    assert.equal(nextCalled, true);
  });

  test('responds 403 listing every accepted role when none match', () => {
    const req = { claims: { realm_access: { roles: ['platform:customer'] } } };
    const res = fakeRes();
    let nextCalled = false;
    requireAnyRealmRole('platform:admin', 'platform:marketing')(req, res, () => {
      nextCalled = true;
    });
    assert.equal(nextCalled, false);
    assert.equal(res.statusCode, 403);
    assert.equal(res.body.message, 'Requires one of: platform:admin, platform:marketing');
  });

  test('responds 403 when req.claims is missing entirely', () => {
    const req = {};
    const res = fakeRes();
    let nextCalled = false;
    requireAnyRealmRole('platform:admin')(req, res, () => {
      nextCalled = true;
    });
    assert.equal(nextCalled, false);
    assert.equal(res.statusCode, 403);
  });
});
